src/Campaigns/Repositories/CampaignRepository.php

```diff
@@ -138,6 +140,50 @@ public function update(Campaign $campaign): void
         Hooks::doAction('givewp_campaign_updated', $campaign);
     }
 
+    /**
+     * @unreleased
+     *
+     * @throws Exception
+     */
+    public function addCampaignForm(Campaign $campaign, DonationForm $donationForm, bool $isDefault = false)
+    {
+        Hooks::doAction('givewp_campaign_form_updating', $campaign, $donationForm, $isDefault);
+
+        DB::query('START TRANSACTION');
+
+        try {
+            // Make sure we'll have only one default form
+            if ($isDefault) {
+                DB::table('give_campaign_forms')
+                    ->where('campaign_id', $campaign->id)
+                    ->update([
+                        'is_default' => false,
+                    ]);
+            }
+
+            $table = DB::prefix('give_campaign_forms');
```

Comment:
My example was more of an example of how to set the default form from one query without the need to do three DB actions. Let's have a simple insert here without fancy things like `ON DUPLICATE KEY UPDATE` because this method should do only one thing, and that's creating a relationship between the campaign and the form. Also, this method is used only when a new form is created which means there shouldn't be duplicate entries. 

BTW. `%s` is used for strings, and all parameters here are integers so it would make sense to use `%d` as a placeholder.

Reply:
About it:

> this method is used only when a new form is created

In the current codebase it's true, but we can use this same method to add a form to the campaign at any time like on [this unit test](https://github.com/impress-org/givewp/blob/feature/forms-list-table-GIVE-1219/tests/Unit/Campaigns/Models/CampaignModelTest.php#L58), so we need a way to make sure that we'll not get a `WordPress database error Duplicate entry...` message when someone tries to insert a form already added before - because of that we need the `ON DUPLICATE KEY UPDATE` statement.  

Besides that, we also need a way to make sure the campaign will have only one default form when someone calls this method - because of that we need the additional query at the top that only gets triggered when `$isDefault` is true.

The placeholder was replaced here: 7fa4f07e752ecc8624916b3707b7cb0c2d7ab741

Reply:
> In the current codebase it's true, but we can use this same method to add a form to the campaign at any time like on this unit test

Unit test should create a new form and not use an existing one. 
Basically, you added all this because there is a problem with data (duplicate meta for formBuilderSettings) on your end. See my previous comment about how we can fix that. 

> because of that we need the additional update action at the top that only gets triggered when $isDefault is true.

I don't have an issue with this part.

Reply:
> Unit test should create a new form and not use an existing one.

Yeah! In the mentioned unit test we are doing exactly that on this line: `$newDefaultForm = DonationForm::factory()->create();`

And then updating the campaign default form using this other one: `give(CampaignRepository::class)->addCampaignForm($campaign, $newDefaultForm, true);`

Now, let's suppose that we call the `addCampaignForm()` method informing the `$newDefaultForm` form again as the parameter in some other place on the code. 

Without the `ON DUPLICATE KEY` statement we'll get `Duplicate entry` error because this form is already on DB. 

> Basically, you added all this because there is a problem with data (duplicate meta for formBuilderSettings) on your end. See my previous comment about how we can fix that.


Nope! I added it thinking about a defensive approach because we can use the `addCampaignForm()` method with a form already added before by mistake and throw an error.

Maybe I'm missing something here, but I can't understand how the `ON DUPLICATE KEY` statement can be a problem. Could you please elaborate more on that?

Reply:
Okay, if you think we should do it like this, then it's okay. I don't want to hold this important PR.

Reply:
> Maybe I'm missing something here, but I can't understand how the ON DUPLICATE KEY statement can be a problem. Could you please elaborate more on that?

It is not a problem, it means that we are doing something wrong. There shouldn't be duplicate forms in this table, and if someone tries to insert a form that already exists, he will get an error from DB because we are using capaign_id and form_id to create the composite key - which is totally fine, and even expected.  You can't do that from UI, only from code. So we are already defensive about it without doing anything extra.

Reply:
Resolved in 3ed2a0713bb6462619b9dbe9187058460cb11b17